### DISCUSSION TITLE
How to cancel a workflow.

### DISCUSSION THREAD

**Discussion by johnwc** [elsa-workflows/elsa-core#3163] [Q&A]
@sfmskywalker  When I create a workflow with a lot of items passed to a `ForEach`, and custom activities within the iterator path that run long processes. I can go into the dashboard and select cancel on the running working and see the status change to cancelled, but it doesn't seem to do anything. After a bit I can refresh the workflow instances page, and the workflow is back to the running status.

**Comment by sfmskywalker** [comment]
Sounds like the Cancel operation gets overridden by active execution of workflows that override the Cancelled status. Will need to look into this together with #3093

**Reply by johnwc** [reply]
Also would be nice to add a property to the context that allows a activity to check if the workflow has been canceled, so that it can handle the cancelation and like `Fault()`; return `Cancelled()` or something similar to tell workflow engine it cancelled and did not run.

**Reply by sfmskywalker** [reply]
That's a great idea 👍🏻

**Reply by truthz03** [reply]
It would be nice to be able to cancel a worklow instance by code because I have some workflows starting with a specific MQTT message but after that the next step is a IF which decides if the message really should be executed. 
If the message shouldn't execute it simple stops with Finished as state. 
It would be nicer to get Cancelled as state otherwise it is hard to find fully executed workflow instances inside the "Workflow Instance" page.
Maybe there should be also a possibility to call Cancelled() with a parameter which prevents the instance from beeing logged inside the "Workflow Instances" page. Otherwise there will be a hugh number of Cancelled instances and only a few Finished in my usecase

**Reply by sfmskywalker** [reply]
@raymonddenhaan Do you think we could get some documentation how to do this?

**Reply by raymonddenhaan** [reply]
To me it seems there are 3 issues described here:

### Cancellation of a workflow ###

I'll make sure to add some documentation on cancellation. At the moment this can be done through the UI and an API call.

### The desire to determine based on logic to finish a workflow with a desired state (in this case cancelled) ###

I can see a use to be able to finish workflows in a way that tells the user it has not run the complete course. I myself would not like to use Cancelled for this state as I would like to be able to tell the difference between a process that has been explicitly cancelled with a manual matter and one that has been set through workflow logic.

### Hiding workflows with a certain state from the workflow instances page ###

We just released a first version of filtering on the workflow instances page and we still have a few improvements planned for this. I do not think we should mark instances in the workflow so that they would not show within the workflow instances list. This to me should be filterable in this list instead.
